version: "3.3"

services:
  mongo:
    image: "mongo:4.0.5"
    container_name: katana-mongo
    restart: always
    volumes:
      - "mongo-datadb:/data/db"
      - "mongo-configdb:/data/configdb"

  zookeeper:
    image: confluentinc/cp-zookeeper:5.3.2
    container_name: katana-zookeeper
    hostname: zookeeper
    ports:
      - "2181:2181"
    environment:
        ZOO_MY_ID: 1
        ZOO_SERVERS: server.1=zookeeper:2888:3888
        ZOOKEEPER_CLIENT_PORT: 2181
    volumes:
      - ./zk-kafka/zookeeper/data:/data
      - ./zk-kafka/zookeeper/datalog:/datalog

  kafka:
    image: confluentinc/cp-enterprise-kafka:5.3.2
    container_name: katana-kafka
    hostname: kafka
    ports:
      - "9092:9092"
    environment:
      KAFKA_ADVERTISED_LISTENERS: LISTENER_DOCKER_INTERNAL://kafka:19092,LISTENER_DOCKER_EXTERNAL://${DOCKER_HOST_IP:-127.0.0.1}:9092
      KAFKA_LISTENER_SECURITY_PROTOCOL_MAP: LISTENER_DOCKER_INTERNAL:PLAINTEXT,LISTENER_DOCKER_EXTERNAL:PLAINTEXT
      KAFKA_INTER_BROKER_LISTENER_NAME: LISTENER_DOCKER_INTERNAL
      KAFKA_ZOOKEEPER_CONNECT: "zookeeper:2181"
      KAFKA_BROKER_ID: 1
      KAFKA_LOG4J_LOGGERS: "kafka.controller=INFO,kafka.producer.async.DefaultEventHandler=INFO,state.change.logger=INFO"
      KAFKA_OFFSETS_TOPIC_REPLICATION_FACTOR: 1
      KAFKA_CREATE_TOPICS: "slice:1:1"
    volumes:
      - ./zk-kafka/kafka/data:/var/lib/kafka/data
    depends_on:
      - zookeeper

  katana-nbi:
    build: ./katana-nbi
    image: katana-nbi
    container_name: katana-nbi
    command: gunicorn -b 0.0.0.0:8000 --workers 3 --access-logfile - --reload "katana.app:create_app()"
    volumes:
      - "./katana-nbi:/katana-nbi"
    ports:
      - "8000:8000"
    environment:
      PYTHONWARNINGS: "ignore:Unverified HTTPS request"
    restart: always
    depends_on:
      - kafka
      - mongo
    # logging:
    #   driver: syslog
    #   options:
    #     syslog-address: "udp://10.30.0.245:5140"
    #     tag: "SliceManager"
    #     syslog-format: "rfc5424"

  katana-mngr:
    build: ./katana-mngr
    image: katana-mngr
    container_name: katana-mngr
    command: python3 katana-mngr.py
    volumes:
      - "./katana-mngr:/katana-mngr"
    environment:
      PYTHONWARNINGS: "ignore:Unverified HTTPS request"
    restart: always
    depends_on:
      - katana-nbi
      - mongo
    # logging:
    #   driver: syslog
    #   options:
    #     syslog-address: "udp://10.30.0.245:5140"
    #     tag: "SliceManager"
    #     syslog-format: "rfc5424"

  katana-cli:
    build: ./katana-cli
    image: katana-cli
    container_name: katana-cli
    command: >
      /bin/bash
    stdin_open: true
    tty: true
    network_mode: "host"
    restart: always
    volumes:
      - "./katana-cli:/cli"
    depends_on:
      - katana-nbi

  swagger:
    image: "swaggerapi/swagger-ui"
    container_name: katana-swagger
    restart: always
    environment:
      - SWAGGER_JSON=/my_swagger/swagger.json
    ports:
      - "8001:8080"
    volumes:
      - "./swagger:/my_swagger"
    depends_on:
      - katana-mngr

  postgres:
    env_file:
      - "./katana-ui/.env"
    image: "postgres:10.3-alpine"
    container_name: katana-postgres
    ports:
      - "5432:5432"
    restart: always
    volumes:
      - "postgres:/var/lib/postgresql/data"

  katana-ui:
    build: ./katana-ui
    image: katana-ui
    container_name: katana-ui
    command: gunicorn -b 0.0.0.0:8080 --workers 3 --access-logfile - --reload "ui.app:create_app()"
    depends_on:
      - "postgres"
    env_file:
      - "./katana-ui/.env"
    ports:
      - "8080:8080"
    restart: always
    volumes:
      - "./katana-ui:/app"
    depends_on:
      - katana-mngr
      - postgres

  # nginx reverse-proxy
  #
  # katana-nginx: 
  #   image: nginx:1.17-alpine
  #   container_name: katana-nginx
  #   depends_on:
  #     - "katana-mngr"
  #     - "katana-ui"
  #   volumes:
  #     - ./katana-nginx/nginx.conf:/etc/nginx/nginx.conf
  #   ports:
  #     - 80:80
  #     - 443:443

volumes:
  mongo-datadb: {}
  mongo-configdb: {}
  postgres: {}
